Add r-varnish

matthewfeickert opened the pull request:
* Add the varnish R package.
   - c.f. https://github.com/carpentries/varnish

Note this recipe was created (note https://github.com/conda/conda-build/issues/5674#issuecomment-2839631202) with

```
conda skeleton cran --use-noarch-generic --add-cross-r-base --no-comments --allow-archived https://github.com/carpentries/varnish
```

and then manually edited to use the correct version given https://github.com/bgruening/conda_r_skeleton_helper/issues/68

<!--
Thank you very much for putting in this recipe PR!

This repository is very active, so if you need help with a PR, please let the
right people know. There are language-specific teams for reviewing recipes.

| Language        | Name of review team           |
| --------------- | ----------------------------- |
| python          | `@conda-forge/help-python`    |
| python/c hybrid | `@conda-forge/help-python-c`  |
| r               | `@conda-forge/help-r`         |
| java            | `@conda-forge/help-java`      |
| nodejs          | `@conda-forge/help-nodejs`    |
| c/c++           | `@conda-forge/help-c-cpp`     |
| perl            | `@conda-forge/help-perl`      |
| Julia           | `@conda-forge/help-julia`     |
| ruby            | `@conda-forge/help-ruby`      |
| Rust            | `@conda-forge/help-rust`      |
| Go              | `@conda-forge/help-go`        |
| other           | `@conda-forge/staged-recipes` |

Once the PR is ready for review, please mention one of the teams above in a
new comment. i.e. `@conda-forge/help-some-language, ready for review!`
Then, a bot will label the PR as 'review-requested'.

Due to GitHub limitations, first time contributors to conda-forge are unable
to ping conda-forge teams directly, but you can [ask a bot to ping the team][1]
using a special command in a comment on the PR to get the attention of the
`staged-recipes` team. You can also consider asking on our [Zulip chat][2]
if your recipe isn't reviewed promptly.

[1]: https://conda-forge.org/docs/maintainer/infrastructure.html#conda-forge-admin-please-ping-team
[2]: https://conda-forge.zulipchat.com

All apologies in advance if your recipe PR does not receive prompt attention.
This is a high volume repository and the reviewers are volunteers. Review times
vary depending on the number of reviewers on a given language team and may be days
or weeks. We are always looking for more staged-recipe reviewers. If you are
interested in volunteering, please contact a member of @conda-forge/core.
We'd love to have the help!
-->

Checklist
- [x] Title of this PR is meaningful: e.g. "Adding my_nifty_package", not "updated meta.yaml".
- [x] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/5eddbd7fc9d1502169089da06c3688d9759be978/recipes/example/meta.yaml#L64-L73) for an example).
- [x] Source is from official source.
- [x] Package does not vendor other packages. (If a package uses the source of another package, they should be separate packages or the licenses of all packages need to be packaged).
- [N/A] If static libraries are linked in, the license of the static library is packaged.
- [x] Package does not ship static libraries. If static libraries are needed, [follow CFEP-18](https://github.com/conda-forge/cfep/blob/main/cfep-18.md).
- [x] Build number is 0.
- [x] A tarball (`url`) rather than a repo (e.g. `git_url`) is used in your recipe (see [here](https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos) for more details).
- [x] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there.
- [x] When in trouble, please check our [knowledge base documentation](https://conda-forge.org/docs/maintainer/knowledge_base.html) before pinging a team.

conda-forge-admin commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipes/r-varnish/meta.yaml```) and found it was in an excellent condition.

matthewfeickert commented:
Ah, this also has JavaScript dependencies given (https://github.com/carpentries/varnish/blob/6b3874a5ee1b14c44cf0375d93ac98d4fea5bd33/README.md?plain=1#L38) and so those should be added as well.

matthewfeickert commented:
https://github.com/carpentries/varnish/blob/6b3874a5ee1b14c44cf0375d93ac98d4fea5bd33/postinstall.sh might prove troublesome, given https://github.com/carpentries/varnish/blob/6b3874a5ee1b14c44cf0375d93ac98d4fea5bd33/package.json requires `feather-icons`, `jquery-visible`, and `svgo` which are JS packages not on conda-forge yet.

matthewfeickert commented:
> https://github.com/carpentries/varnish/blob/6b3874a5ee1b14c44cf0375d93ac98d4fea5bd33/postinstall.sh might prove troublesome, given https://github.com/carpentries/varnish/blob/6b3874a5ee1b14c44cf0375d93ac98d4fea5bd33/package.json requires `feather-icons`, `jquery-visible`, and `svgo` which are JS packages not on conda-forge yet.

@froggleston @ErinBecker @astroDimitrios can you confirm that for `varnish` to work as expected that the [all the JavaScript dependencies given in `package.json`](https://github.com/carpentries/varnish/blob/1eabe9c5f2f98b5407303db9ff446b761c7ba173/package.json#L11-L17)

```
  "dependencies": {
    "bootstrap": "5.3.2",
    "feather-icons": "4.28.0",
    "jquery": "3.6.0",
    "jquery-visible": "^1.2.0",
    "svgo": "^2.8.0"
  },
```

also need to get installed? From looking at https://github.com/carpentries/varnish/'s CI I don't fully understand how the minification of the JS is happening.

matthewfeickert commented:
> can you confirm that for `varnish` to work as expected that the [all the JavaScript dependencies given in `package.json`](https://github.com/carpentries/varnish/blob/1eabe9c5f2f98b5407303db9ff446b761c7ba173/package.json#L11-L17)
> 
> ```
>   "dependencies": {
>     "bootstrap": "5.3.2",
>     "feather-icons": "4.28.0",
>     "jquery": "3.6.0",
>     "jquery-visible": "^1.2.0",
>     "svgo": "^2.8.0"
>   },
> ```
> 
> also need to get installed? From looking at https://github.com/carpentries/varnish/'s CI I don't fully understand how the minification of the JS is happening in terms of distribution.

Apparently not given that installing `varnish` from https://carpentries.r-universe.dev produces the same `lib/R/library/varnish/pkgdown/` directory structure and files as just building the package on conda-forge without additional JavaScript dependencies. So that's good.

<details>
<summary>carpentries.r-universe.dev varnish install records:</summary>

```console
$ docker run --rm -ti ghcr.io/prefix-dev/pixi:latest
root@842141a41e6d:/# pixi init example
Created /example/pixi.toml
root@842141a41e6d:/# cd example/
root@842141a41e6d:/example# pixi add r-base
Added r-base >=4.4.3,<4.5
root@842141a41e6d:/example# pixi run Rscript -e 'install.packages("varnish", repos = "https://carpentries.r-universe.dev")'
trying URL 'https://carpentries.r-universe.dev/src/contrib/varnish_1.0.5.tar.gz'
Content type 'application/gzip' length 6721847 bytes (6.4 MB)
==================================================
downloaded 6.4 MB

* installing *source* package 'varnish' ...
** using staged installation
** R
** inst
** byte-compile and prepare package for lazy loading
** help
*** installing help indices
*** copying figures
** building package indices
** testing if installed package can be loaded from temporary location
** testing if installed package can be loaded from final location
** testing if installed package keeps a record of temporary installation path
* DONE (varnish)

The downloaded source packages are in
	'/tmp/RtmphA5sXv/downloaded_packages'
Updating HTML index of packages in '.Library'
Making 'packages.html' ... done
root@842141a41e6d:/example# pixi global install tree
Global environments as specified in '/root/.pixi/manifests/pixi-global.toml'
└── tree: 2.2.1 (installed)
    └─ exposes: tree
root@842141a41e6d:/example# tree -L 2 .pixi/envs/default/lib/R/library/varnish/pkgdown/
.pixi/envs/default/lib/R/library/varnish/pkgdown/
|-- assets
|   |-- android-chrome-192x192.png
|   |-- android-chrome-512x512.png
|   |-- apple-touch-icon.png
|   |-- assets
|   |-- favicon-16x16.png
|   |-- favicon-32x32.png
|   |-- favicons
|   |-- mstile-150x150.png
|   |-- safari-pinned-tab.svg
|   `-- site.webmanifest
`-- templates
    |-- config-docsearch.json
    |-- content-chapter.html
    |-- content-extra.html
    |-- content-overview.html
    |-- content-syllabus.html
    |-- docsearch.html
    |-- footer.html
    |-- head.html
    |-- header.html
    |-- layout.html
    `-- navbar.html

5 directories, 19 files
# tree -L 2 .pixi/envs/default/lib/R/library/varnish/pkgdown/assets/assets/
.pixi/envs/default/lib/R/library/varnish/pkgdown/assets/assets/
|-- fonts
|   |-- Mulish-Black.eot
|   |-- Mulish-Black.svg
|   |-- Mulish-Black.ttf
|   |-- Mulish-Black.woff
|   |-- Mulish-Black.woff2
|   |-- Mulish-BlackItalic.eot
|   |-- Mulish-BlackItalic.svg
|   |-- Mulish-BlackItalic.ttf
|   |-- Mulish-BlackItalic.woff
|   |-- Mulish-BlackItalic.woff2
|   |-- Mulish-Bold.eot
|   |-- Mulish-Bold.svg
|   |-- Mulish-Bold.ttf
|   |-- Mulish-Bold.woff
|   |-- Mulish-Bold.woff2
|   |-- Mulish-BoldItalic.eot
|   |-- Mulish-BoldItalic.svg
|   |-- Mulish-BoldItalic.ttf
|   |-- Mulish-BoldItalic.woff
|   |-- Mulish-BoldItalic.woff2
|   |-- Mulish-ExtraBold.eot
|   |-- Mulish-ExtraBold.svg
|   |-- Mulish-ExtraBold.ttf
|   |-- Mulish-ExtraBold.woff
|   |-- Mulish-ExtraBold.woff2
|   |-- Mulish-ExtraBoldItalic.eot
|   |-- Mulish-ExtraBoldItalic.svg
|   |-- Mulish-ExtraBoldItalic.ttf
|   |-- Mulish-ExtraBoldItalic.woff
|   |-- Mulish-ExtraBoldItalic.woff2
|   |-- Mulish-ExtraLight.eot
|   |-- Mulish-ExtraLight.svg
|   |-- Mulish-ExtraLight.ttf
|   |-- Mulish-ExtraLight.woff
|   |-- Mulish-ExtraLight.woff2
|   |-- Mulish-ExtraLightItalic.eot
|   |-- Mulish-ExtraLightItalic.svg
|   |-- Mulish-ExtraLightItalic.ttf
|   |-- Mulish-ExtraLightItalic.woff
|   |-- Mulish-ExtraLightItalic.woff2
|   |-- Mulish-Italic-VariableFont_wght.ttf
|   |-- Mulish-Italic.eot
|   |-- Mulish-Italic.svg
|   |-- Mulish-Italic.ttf
|   |-- Mulish-Italic.woff
|   |-- Mulish-Italic.woff2
|   |-- Mulish-Light.eot
|   |-- Mulish-Light.svg
|   |-- Mulish-Light.ttf
|   |-- Mulish-Light.woff
|   |-- Mulish-Light.woff2
|   |-- Mulish-LightItalic.eot
|   |-- Mulish-LightItalic.svg
|   |-- Mulish-LightItalic.ttf
|   |-- Mulish-LightItalic.woff
|   |-- Mulish-LightItalic.woff2
|   |-- Mulish-Medium.eot
|   |-- Mulish-Medium.svg
|   |-- Mulish-Medium.ttf
|   |-- Mulish-Medium.woff
|   |-- Mulish-Medium.woff2
|   |-- Mulish-MediumItalic.eot
|   |-- Mulish-MediumItalic.svg
|   |-- Mulish-MediumItalic.ttf
|   |-- Mulish-MediumItalic.woff
|   |-- Mulish-MediumItalic.woff2
|   |-- Mulish-Regular.eot
|   |-- Mulish-Regular.svg
|   |-- Mulish-Regular.ttf
|   |-- Mulish-Regular.woff
|   |-- Mulish-Regular.woff2
|   |-- Mulish-SemiBold.eot
|   |-- Mulish-SemiBold.svg
|   |-- Mulish-SemiBold.ttf
|   |-- Mulish-SemiBold.woff
|   |-- Mulish-SemiBold.woff2
|   |-- Mulish-SemiBoldItalic.eot
|   |-- Mulish-SemiBoldItalic.svg
|   |-- Mulish-SemiBoldItalic.ttf
|   |-- Mulish-SemiBoldItalic.woff
|   |-- Mulish-SemiBoldItalic.woff2
|   |-- Mulish-VariableFont_wght.ttf
|   |-- MulishExtraLight-Regular.eot
|   |-- MulishExtraLight-Regular.svg
|   |-- MulishExtraLight-Regular.woff
|   |-- MulishExtraLight-Regular.woff2
|   |-- mulish-v5-latin-regular.eot
|   |-- mulish-v5-latin-regular.svg
|   |-- mulish-v5-latin-regular.ttf
|   |-- mulish-v5-latin-regular.woff
|   |-- mulish-v5-latin-regular.woff2
|   |-- mulish-variablefont_wght.woff
|   `-- mulish-variablefont_wght.woff2
|-- images
|   |-- carpentries-logo-sm.svg
|   |-- carpentries-logo.svg
|   |-- data-logo-sm.svg
|   |-- data-logo.svg
|   |-- dropdown-arrow.svg
|   |-- incubator-logo-sm.svg
|   |-- incubator-logo.svg
|   |-- lab-logo-sm.svg
|   |-- lab-logo.svg
|   |-- library-logo-sm.svg
|   |-- library-logo.svg
|   |-- minus.svg
|   |-- plus.svg
|   |-- software-logo-sm.svg
|   `-- software-logo.svg
|-- scripts.js
|-- styles.css
|-- styles.css.map
`-- themetoggle.js

3 directories, 112 files
```

</details>

@froggleston @ErinBecker @astroDimitrios if any of you would like to be maintainers on the `r-varnish` feedstock (no work expected from you) I'd be happy to add you. Whenever I add something to conda-forge I like to alert the maintainers so that they can still have control over the distributions of the software they create.

matthewfeickert commented on recipes/r-varnish/meta.yaml:
```diff
@@ -0,0 +1,56 @@
+    - {{ posix }}filesystem        # [win]
+    - {{ posix }}git
+    - {{ posix }}zip               # [win]
+    - cross-r-base {{ r_base }}    # [build_platform != target_platform]
```
If this is `noarch` is this needed at all?

matthewfeickert commented:
@conda-forge/help-r, ready for review!

astroDimitrios commented:
> > can you confirm that for `varnish` to work as expected that the [all the JavaScript dependencies given in `package.json`](https://github.com/carpentries/varnish/blob/1eabe9c5f2f98b5407303db9ff446b761c7ba173/package.json#L11-L17)
> > ```
> >   "dependencies": {
> >     "bootstrap": "5.3.2",
> >     "feather-icons": "4.28.0",
> >     "jquery": "3.6.0",
> >     "jquery-visible": "^1.2.0",
> >     "svgo": "^2.8.0"
> >   },
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > also need to get installed? From looking at https://github.com/carpentries/varnish/'s CI I don't fully understand how the minification of the JS is happening in terms of distribution.
> 
> Apparently not given that installing `varnish` from https://carpentries.r-universe.dev produces the same `lib/R/library/varnish/pkgdown/` directory structure and files as just building the package on conda-forge without additional JavaScript dependencies. So that's good.
> carpentries.r-universe.dev varnish install records:
> 
> @froggleston @ErinBecker @astroDimitrios if any of you would like to be maintainers on the `r-varnish` feedstock (no work expected from you) I'd be happy to add you. Whenever I add something to conda-forge I like to alert the maintainers so that they can still have control over the distributions of the software they create.

I'm not formally a maintainer but have done some JS work on Varnish.
I think as long as the squashed js css etc files are included by the conda install (for instance https://github.com/carpentries/varnish/blob/main/inst/pkgdown/assets/assets/scripts.js) then it will work. Installing the js libraries with npm, postinstall, squashing the sass and js are all for developers/modifiers of Varnish (at least that's my understanding).

froggleston commented:
@astroDimitrios is correct - the minification and squashing happens [via CI](https://github.com/carpentries/varnish/blob/main/.github/workflows/compile-css.yaml), but manually squashing/minifying assets will work fine via the same scripts provided with varnish.

matthewfeickert commented:
@froggleston @astroDimitrios Thanks for the info. I just wanted to make sure that there weren't additional steps that required JavaScript packages to be installed to build the package. :+1: 

> The functionality of the package isn't tested in the recipe. Have you [downloaded an artifact](https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=1236039&view=artifacts&pathAsName=false&type=publishedArtifacts) to manually check it works in your expected use case?
> 
> If you're confident with the functionality, then I'll merge.

@mfansler Thanks for your fast review. I don't know how to test `r-vanish` by itself, as the docs don't provide any context (https://carpentries.github.io/varnish/) and strongly guide you to just view `varnish` as a dependency of `sandpaper` and ignore it. So unless @froggleston has advice here I'd say go for merge. :+1:

mfansler commented:
>  I don't know how to test r-vanish by itself, as the docs don't provide any context (https://carpentries.github.io/varnish/) and strongly guide you to just view varnish as a dependency of sandpaper and ignore it.

In that case how about also adding `r-sandpaper` in this PR and running its unit tests?

matthewfeickert commented:
> In that case how about also adding `r-sandpaper` in this PR and running its unit tests?

`r-sandpaper` doesn't exist yet, as it has requirements on `r-tinkr`, `r-varnish`, and `r-pegboard` (https://github.com/conda-forge/staged-recipes/pull/29889) and so needs this PR to go in first.

`r-varnish` doesn't have any `testhat` tests given that it 

> [does not contain any useful R code. It contains HTML templates with mustache syntax, CSS, and JavaScript.](https://github.com/carpentries/varnish/blob/e7d332c18f604ae9927d7857328ae261e7a624c5/_pkgdown.yml#L5-L6)

so if I understand @froggleston there's nothing to test other than if the template files exist.

mfansler commented:
> r-sandpaper doesn't exist yet, as it has requirements on r-tinkr, r-varnish, and r-pegboard (#29889) and so needs this PR to go in first.

If you add it in this PR, the DAG will be correctly computed and it will use the local build of r-varnish here to satisfy the requirement at build/test time. Probably just need to wait for r-pegboard to finish uploading and get indexed on Anaconda Cloud.

matthewfeickert commented on recipes/r-sandpaper/meta.yaml:
```diff
@@ -0,0 +1,119 @@
+test:
+  requires:
+    - r-testthat
+    - r-mockr
```
Oh. `r-mockr` isn't on conda-forge yet, so will need https://github.com/krlmlr/mockr/ to get added first.

Putting this back into draft then.

matthewfeickert replied:
c.f. https://github.com/conda-forge/staged-recipes/pull/29899

matthewfeickert commented:
Actually, even with a local cached copy of `r-mockr` built there's issues with `sandpaper`

<details>
<summary>Failed tests summary:</summary>

```pytb
══ Failed tests ════════════════════════════════════════════════════════════════
── Failure (test-ci_session_info.R:4:3): ci_session_info() produces output [plain] ──
`ci_session_info()` produced no output
Backtrace:
     ▆
  1. ├─base::suppressMessages(...) at test-ci_session_info.R:3:3
  2. │ └─base::withCallingHandlers(...)
  3. ├─expect_output(ci_session_info(), "sandpaper") %>% ... at test-ci_session_info.R:4:3
  4. ├─testthat::expect_message(., "Package Information")
  5. │ └─testthat:::expect_condition_matching(...)
  6. │   └─testthat:::quasi_capture(...)
  7. │     ├─testthat (local) .capture(...)
  8. │     │ └─base::withCallingHandlers(...)
  9. │     └─rlang::eval_bare(quo_get_expr(.quo), quo_get_env(.quo))
 10. └─testthat::expect_output(ci_session_info(), "sandpaper")
── Failure (test-ci_session_info.R:4:3): ci_session_info() produces output [plain] ──
`.` did not throw the expected message.
Backtrace:
    ▆
 1. ├─base::suppressMessages(...) at test-ci_session_info.R:3:3
 2. │ └─base::withCallingHandlers(...)
 3. ├─expect_output(ci_session_info(), "sandpaper") %>% ... at test-ci_session_info.R:4:3
 4. └─testthat::expect_message(., "Package Information")
── Failure (test-ci_session_info.R:4:3): ci_session_info() produces output [ansi] ──
`ci_session_info()` produced no output
Backtrace:
     ▆
  1. ├─base::suppressMessages(...) at test-ci_session_info.R:3:3
  2. │ └─base::withCallingHandlers(...)
  3. ├─expect_output(ci_session_info(), "sandpaper") %>% ... at test-ci_session_info.R:4:3
  4. ├─testthat::expect_message(., "Package Information")
  5. │ └─testthat:::expect_condition_matching(...)
  6. │   └─testthat:::quasi_capture(...)
  7. │     ├─testthat (local) .capture(...)
  8. │     │ └─base::withCallingHandlers(...)
  9. │     └─rlang::eval_bare(quo_get_expr(.quo), quo_get_env(.quo))
 10. └─testthat::expect_output(ci_session_info(), "sandpaper")
── Failure (test-ci_session_info.R:4:3): ci_session_info() produces output [ansi] ──
`.` did not throw the expected message.
Backtrace:
    ▆
 1. ├─base::suppressMessages(...) at test-ci_session_info.R:3:3
 2. │ └─base::withCallingHandlers(...)
 3. ├─expect_output(ci_session_info(), "sandpaper") %>% ... at test-ci_session_info.R:4:3
 4. └─testthat::expect_message(., "Package Information")
── Failure (test-ci_session_info.R:4:3): ci_session_info() produces output [unicode] ──
`ci_session_info()` produced no output
Backtrace:
     ▆
  1. ├─base::suppressMessages(...) at test-ci_session_info.R:3:3
  2. │ └─base::withCallingHandlers(...)
  3. ├─expect_output(ci_session_info(), "sandpaper") %>% ... at test-ci_session_info.R:4:3
  4. ├─testthat::expect_message(., "Package Information")
  5. │ └─testthat:::expect_condition_matching(...)
  6. │   └─testthat:::quasi_capture(...)
  7. │     ├─testthat (local) .capture(...)
  8. │     │ └─base::withCallingHandlers(...)
  9. │     └─rlang::eval_bare(quo_get_expr(.quo), quo_get_env(.quo))
 10. └─testthat::expect_output(ci_session_info(), "sandpaper")
── Failure (test-ci_session_info.R:4:3): ci_session_info() produces output [unicode] ──
`.` did not throw the expected message.
Backtrace:
    ▆
 1. ├─base::suppressMessages(...) at test-ci_session_info.R:3:3
 2. │ └─base::withCallingHandlers(...)
 3. ├─expect_output(ci_session_info(), "sandpaper") %>% ... at test-ci_session_info.R:4:3
 4. └─testthat::expect_message(., "Package Information")
── Failure (test-ci_session_info.R:4:3): ci_session_info() produces output [fancy] ──
`ci_session_info()` produced no output
Backtrace:
     ▆
  1. ├─base::suppressMessages(...) at test-ci_session_info.R:3:3
  2. │ └─base::withCallingHandlers(...)
  3. ├─expect_output(ci_session_info(), "sandpaper") %>% ... at test-ci_session_info.R:4:3
  4. ├─testthat::expect_message(., "Package Information")
  5. │ └─testthat:::expect_condition_matching(...)
  6. │   └─testthat:::quasi_capture(...)
  7. │     ├─testthat (local) .capture(...)
  8. │     │ └─base::withCallingHandlers(...)
  9. │     └─rlang::eval_bare(quo_get_expr(.quo), quo_get_env(.quo))
 10. └─testthat::expect_output(ci_session_info(), "sandpaper")
── Failure (test-ci_session_info.R:4:3): ci_session_info() produces output [fancy] ──
`.` did not throw the expected message.
Backtrace:
    ▆
 1. ├─base::suppressMessages(...) at test-ci_session_info.R:3:3
 2. │ └─base::withCallingHandlers(...)
 3. ├─expect_output(ci_session_info(), "sandpaper") %>% ... at test-ci_session_info.R:4:3
 4. └─testthat::expect_message(., "Package Information")
── Failure (test-overview.R:32:3): We can switch between overview and regular lesson metadata ──
lmeta$url does not match "/lesson-example/".
Actual value: "https://FIXME-template-source.github.io/NA/"
Backtrace:
    ▆
 1. └─testthat::expect_match(lmeta$url, "/lesson-example/", fixed = TRUE) at test-overview.R:32:3
 2.   └─testthat:::expect_match_(...)
── Failure (test-overview.R:67:3): We can switch between overview and regular lesson metadata ──
lmeta$url does not match "/lesson-example/".
Actual value: "https://FIXME-template-source.github.io/NA/"
Backtrace:
    ▆
 1. └─testthat::expect_match(lmeta$url, "/lesson-example/", fixed = TRUE) at test-overview.R:67:3
 2.   └─testthat:::expect_match_(...)

[ FAIL 10 | WARN 2 | SKIP 108 | PASS 890 ]
Deleting unused snapshots:
• utils-git/diff.md

ℹ removing 'sandpaper-local' (/tmp/Rtmp71zYkg/REMOTE-5796ba53317)
[ FAIL 1 | WARN 0 | SKIP 0 | PASS 0 ]

── Error (testthat.R:4:1): (code run outside of `test_that()`) ─────────────────
Error: Test failures
Backtrace:
    ▆
 1. └─testthat::test_check("sandpaper") at testthat.R:4:1
 2.   └─testthat::test_dir(...)
 3.     └─testthat:::test_files(...)
 4.       └─testthat:::test_files_serial(...)
 5.         └─testthat:::test_files_check(...)

[ FAIL 1 | WARN 0 | SKIP 0 | PASS 0 ]

```

</details>

@froggleston, sorry to bug you, but do you have any thoughts here?

danielnachun commented:
FYI it will not really be possible to package those NodeJS packages as a separate package.  You'll want to basically use code like this to install the modules in `build.sh`:
```
# Create package archive and install globally
npm pack --ignore-scripts
npm install -ddd \
    --global \
    --build-from-source \
    ${SRC_DIR}/the-carpentries-${PKG_VERSION}.tgz

# Create license report for dependencies
pnpm install
pnpm-licenses generate-disclaimer --prod --output-file=third-party-licenses.txt
```
You'll also need to add `nodejs`, `pnpm` and `pnpm-licenses` to the `build` requirements and `third-party-licenses.txt` to the `license_file` entry.

matthewfeickert commented:
> FYI it will not really be possible to package those NodeJS packages as a separate package. You'll want to basically use code like this to install the modules in `build.sh`:

@danielnachun luckily we don't have to given info in https://github.com/conda-forge/staged-recipes/pull/29866#issuecomment-2842753029. I was confused on what the upstream CI was actually doing, but those are JavaScript packages are just used in the upstream CI to minify everything into [`varnish/pkgdown/assets/assets/scripts.js`](https://github.com/carpentries/varnish/blob/e7d332c18f604ae9927d7857328ae261e7a624c5/inst/pkgdown/assets/assets/scripts.js). So there's no additional steps that need to be taken.

matthewfeickert commented:
@danielnachun oh, or are you saying that the minification as it is now is actually breaking the license and so `varnish` would need to update its license to be compliant with conda-forge requirements?

froggleston commented:
> @froggleston, sorry to bug you, but do you have any thoughts here?

I *think* the `test-ci_session_info` issue is that the sessioninfo package needs to be available.

As for the URL mismatch, I think that's a problem with how the test is being called and the config environment being set up. I'll do some tests here and report back!

matthewfeickert commented:
> I _think_ the `test-ci_session_info` issue is that the sessioninfo package needs to be available.
> 
> As for the URL mismatch, I think that's a problem with how the test is being called and the config environment being set up. I'll do some tests here and report back!

Thanks @froggleston. Adding in `r-sessioninfo` does drop the number of errors to just two

```
══ Failed tests ════════════════════════════════════════════════════════════════
── Failure (test-overview.R:32:3): We can switch between overview and regular lesson metadata ──
lmeta$url does not match "/lesson-example/".
Actual value: "https://FIXME-template-source.github.io/NA/"
Backtrace:
    ▆
 1. └─testthat::expect_match(lmeta$url, "/lesson-example/", fixed = TRUE) at test-overview.R:32:3
 2.   └─testthat:::expect_match_(...)
── Failure (test-overview.R:67:3): We can switch between overview and regular lesson metadata ──
lmeta$url does not match "/lesson-example/".
Actual value: "https://FIXME-template-source.github.io/NA/"
Backtrace:
    ▆
 1. └─testthat::expect_match(lmeta$url, "/lesson-example/", fixed = TRUE) at test-overview.R:67:3
 2.   └─testthat:::expect_match_(...)

[ FAIL 2 | WARN 2 | SKIP 108 | PASS 898 ]
Deleting unused snapshots:
• utils-git/diff.md

ℹ removing 'sandpaper-local' (/tmp/Rtmppz45g1/REMOTE-58a7c20b3ae)
[ FAIL 1 | WARN 0 | SKIP 0 | PASS 0 ]

── Error (testthat.R:4:1): (code run outside of `test_that()`) ─────────────────
Error: Test failures
Backtrace:
    ▆
 1. └─testthat::test_check("sandpaper") at testthat.R:4:1
 2.   └─testthat::test_dir(...)
 3.     └─testthat:::test_files(...)
 4.       └─testthat:::test_files_serial(...)
 5.         └─testthat:::test_files_check(...)

[ FAIL 1 | WARN 0 | SKIP 0 | PASS 0 ]
Error: Test failures
Execution halted
```

> Update: I think I found the URL issue with the tests and should be fixed in this PR: [carpentries/sandpaper#651](https://github.com/carpentries/sandpaper/pull/651). This will be in the upcoming release of sandpaper 0.16.12.

Great! Thanks a bunch, @froggleston.

@mfansler Given this, and that the `r-varnish` related bits of the `r-sandpaper` tests are passing, and that we're waiting on `GPL-3.0-only` vs. `GPL-3.0-or-later` clarification on PR https://github.com/conda-forge/staged-recipes/pull/29899, I'm going to move `sandpaper` to its own PR and follow up there once all of the `sandpaper` requirements and fixes are met.

So this PR will go back to just being `r-varnish`, and can be merged whenever, unless @danielnachun has clarification on any concerns about minified JavaScript being included in `r-varnish`.

danielnachun commented:
Thank you for clarifying what the NodeJS packages are actually used for.  If they don't need to be shipped in the package, they don't need to have their licenses bundled.  Third party license concerns only come into play for language ecosystems like NodeJS, JDK family or .NET family languages that always bundle all of their dependencies, or languages like Go, Rust, Haskell or OCaml which statically link their dependencies into a single executable.  Since neither one of these apply here, I believe the license information you already have is sufficient for this package.

matthewfeickert commented:
> Since neither one of these apply here, I believe the license information you already have is sufficient for this package.

Yup, I think we're all in agreement now. Thanks very much for thoughtfully stepping through this, @danielnachun. :+1:
